### PR TITLE
Use `pathlib.Path` where possible. 

### LINKED ISSUE #637: Use pathlib
This is actually a bit involved. It would be best to start by tackling 1-2 functions/methods at a time. Find a function/method that uses a file path string and make it be able to handle Path or str objects, for example.

### PR DESCRIPTION
**Pull request by desilinguist**
- Update all code to use `pathlib.Path` instead of `os.path.*` where possible. 
- Add `pyproject.toml`. 
- Fix docstrings to pass pydocstyle pre-commit checks.
- Update `.pre-commit-config.yaml` to specify more flags for `ruff` so that it can cover pydocstyle, isort, and flake8 checks. 
- Other minor changes: replace unused variables with `_` etc. 

This PR closes #637.

### CODE REVIEW

File: skll/config/__init__.py [hunk 1]
```diff
@@ -1016,9 +1019,13 @@ def _setup_config_parser(config_path, validate=True):
     # initialize config parser with the given defaults
```
Comment (tazin-afrin) [inline skll/config/__init__.py:1019]: Raises FileNotFoundError instead of IOError.

File: skll/data/readers.py [hunk 2]
```diff
@@ -1052,8 +1077,9 @@ def _sub_read(self, file):
 
 def safe_float(text, replace_dict=None, logger=None):
     """
-    Attempts to convert a string to an int, and then a float, but if neither is
-    possible, returns the original string value.
+    Convert string to int and then float.
```
Comment (tazin-afrin) [inline skll/data/readers.py:1080]:
```suggestion
    Convert string to int.
    If not possible,  then float.
```

Comment (desilinguist) [reply skll/data/readers.py:1080]: That's not actually correct. It does do both.

Comment (desilinguist) [reply skll/data/readers.py:1080]: Good catch. I'll tweak it to be more accurate.

File: skll/config/__init__.py [hunk 3]
```diff
@@ -1016,9 +1019,13 @@ def _setup_config_parser(config_path, validate=True):
...
     config = SKLLConfigParser()
 
+    # convert config_path to a Path object
+    config_path = Path(config_path)
+
     # Read file if it exists
-    if not exists(config_path):
-        raise IOError(errno.ENOENT, "Configuration file does not exist", config_path)
+    if not config_path.exists():
+        raise FileNotFoundError("The configuration file {config_path} does not exist.")
```
[the remark this answers is not in the document]
Comment (desilinguist) [reply skll/config/__init__.py:1027]: Good catch!

File: skll/learner/__init__.py [hunk 4]
```diff
@@ -1737,7 +1732,9 @@ def learning_curve(
         override_minimum=False,
     ):
         """
-        Generates learning curves for a given model on the training examples
+        Generate learning curves for the learner using the examples.
+
+        The learning curves arge generated on the training examples
```
Comment (tamarl08) [inline skll/learner/__init__.py:1737]:
```suggestion
        The learning curves are generated on the training examples
```

File: skll/learner/utils.py [hunk 5]
```diff
@@ -496,8 +506,7 @@ def contiguous_ints_or_floats(numbers):
 
 def get_acceptable_classification_metrics(label_array):
     """
-    Return the set of metrics that are acceptable given the
-    the unique set of labels that we are classifying.
+    Return acceptable metrics given the the unique set of labels being classified.
```
Comment (tamarl08) [inline skll/learner/utils.py:509]:
```suggestion
    Return acceptable metrics given the unique set of labels being classified.
```

File: skll/learner/utils.py [hunk 6]
```diff
@@ -447,6 +456,8 @@ def compute_num_folds_from_example_counts(cv_folds, labels, model_type, logger=N
 
 def contiguous_ints_or_floats(numbers):
     """
+    Check for continuity in given list of numbers.
```
Comment (tamarl08) [inline skll/learner/utils.py:459]:
```suggestion
    Check for continuity in the given list of numbers.
```

File: skll/utils/commandline/plot_learning_curves.py [hunk 7]
```diff
@@ -52,18 +49,22 @@ def main(argv=None):
     logging.captureWarnings(True)
     logging.basicConfig(format="%(asctime)s - %(name)s - %(levelname)s - " "%(message)s")
 
+    # conver to Path objects
```
Comment (tamarl08) [inline skll/utils/commandline/plot_learning_curves.py:52]:
```suggestion
    # convert to Path objects
```

**Review by tamarl08 — APPROVED** [review]

**Pull request merged**